jobs/kpi-forecasting/kpi_forecasting.py

```diff
@@ -11,22 +25,160 @@
+            self.segments = list(self.config_data["metric_hub"]["segments"].keys())
+        elif model_type == "prophet":
+            self.model_type = "prophet"
+            self.model_class = ProphetForecast
+            self.segments = None
+        else:
+            raise ValueError(f"Model type {model_type} is not supported")
+
+    def add_metadata(self, summary_df):
+        # add Metric Hub metadata columns
+        summary_df["metric_alias"] = self.metric_hub.alias.lower()
+        summary_df["metric_hub_app_name"] = self.metric_hub.app_name.lower()
+        summary_df["metric_hub_slug"] = self.metric_hub.slug.lower()
+        summary_df["metric_start_date"] = pd.to_datetime(self.metric_hub.min_date)
+        summary_df["metric_end_date"] = pd.to_datetime(self.metric_hub.max_date)
+        summary_df["metric_collected_at"] = self.collected_at
+
+        # add forecast model metadata columns
+        summary_df["forecast_start_date"] = self.start_date
+        summary_df["forecast_end_date"] = self.end_date
+        summary_df["forecast_trained_at"] = self.trained_at
+        summary_df["forecast_predicted_at"] = self.predicted_at
+
+        return summary_df
+
+    def get_raw_data(self):
+        metric_hub = MetricHub(**self.config_data["metric_hub"])
+        self.collected_at = datetime.now(timezone.utc).replace(tzinfo=None)
+        observed_df = metric_hub.fetch()
+        # set attribute to generate metadata later
+        self.metric_hub = metric_hub
+        return observed_df
+
+    def get_predict_dates(self, observed_df):
+        start_date = pd.to_datetime(
+            self.config_data["forecast_model"]["forecast_start"]
+            or self._default_start_date(observed_df)
+        )
+        self.start_date = start_date
+        end_date = pd.to_datetime(
+            self.config_data["forecast_model"]["forecast_end"]
+            or self._default_end_date()
+        )
+        self.end_date = end_date
+        return pd.DataFrame(
+            {"submission_date": pd.date_range(start_date, end_date).date}
+        )
+
+    def fit(self, observed_df):
+        # the model parameters are mixed in a bit with paraters to configure
+        # the predict dates so we have to do this
+
+        if self.model_type == "funnel":
+            model_parameters = {
+                "parameters": self.config_data["forecast_model"]["parameters"]
+            }
+            model_parameters["segments"] = self.segments
+        elif self.model_type == "prophet":
+            model_parameters = self.config_data["forecast_model"]["parameters"]
+        model = self.model_class(**model_parameters)
+        self.trained_at = datetime.now(timezone.utc).replace(tzinfo=None)
+        return model.fit(observed_df)
+
+    def predict_and_summarize(self, model, predict_dates, observed_df):
+        raw_predictions = model.predict(predict_dates)
+        self.predicted_at = datetime.now(timezone.utc).replace(tzinfo=None)
```

Comment:
Just to be consistent with the `fit` class, which sets the timestamp immediately before fitting.
```suggestion
        self.predicted_at = datetime.now(timezone.utc).replace(tzinfo=None)
        raw_predictions = model.predict(predict_dates)
```